p2p/peer.go

```diff
@@ -310,28 +291,52 @@ func (p *peer) OnStart() error {
 		return err
 	}
 
-	if err := p.mconn.Start(); err != nil {
-		return err
-	}
-
+	go p.processMessages()
 	go p.metricsReporter()
+
 	return nil
 }
 
+// processMessages processes messages received from the connection.
+func (p *peer) processMessages() {
+	defer func() {
+		if r := recover(); r != nil {
+			p.Logger.Error("Peer message processing panic", "err", r, "stack", string(debug.Stack()))
+			p.onError(fmt.Errorf("panic during peer message processing: %v", r))
+		}
+	}()
+
+	for {
+		chID, msg, err := p.conn.ReceiveMessage()
+		if err != nil {
+			p.onError(err)
+			return
+		}
+		reactor, ok := p.reactors[chID]
+		if !ok {
+			p.onError(fmt.Errorf("unknown channel %v", chID))
+			return
+		}
+		reactor.Receive(chID, p, msg)
+	}
+}
+
 // FlushStop mimics OnStop but additionally ensures that all successful
 // .Send() calls will get flushed before closing the connection.
 // NOTE: it is not safe to call this method more than once.
 func (p *peer) FlushStop() {
 	p.metricsTicker.Stop()
 	p.BaseService.OnStop()
-	p.mconn.FlushStop() // stop everything and close the conn
+	if err := p.conn.FlushClose(); err != nil {
+		p.Logger.Debug("Error while stopping peer", "err", err)
```

Comment:
[style-nit]

```suggestion
		p.Logger.Debug("error while stopping peer", "err", err)
```

Reply:
We currently capitalize all log messages. We could change that, but in that case I think we should do it everywhere rather than having inconsistent logs.

Reply:
In all my PR's I've been making this change. We can do it piecemeal or we can do it in one fell swoop. I don't have a preference.

Reply:
Ok, as long as we're planning on migrating all log messages before 0.35 that works for me.

Reply:
Yes, I'll be sure this is taken care of :)

Reply:
This is the sort of thing we should have a linter for. We've discussed other candidates for a custom linter as well (I think unchecked arithmetic was one). Looks like there's some official infrastructure for this, might be a fun project: https://pkg.go.dev/golang.org/x/tools/go/analysis